src/libos/src/fs/file_ops/utimes.rs

```diff
@@ -0,0 +1,154 @@
+use crate::util::mem_util::from_user;
+
+use super::time::{timespec_t, OcclumTimeProvider};
+use super::*;
+
+use rcore_fs::dev::TimeProvider;
+
+const UTIME_NOW: i64 = (1i64 << 30) - 1i64;
+pub const UTIME_OMIT: i64 = (1i64 << 30) - 2i64;
+
+bitflags! {
+    pub struct UtimeFlags: i32 {
+        const AT_SYMLINK_NOFOLLOW = 1 << 8;
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+pub struct utimbuf_t {
+    atime: time_t,
+    mtime: time_t,
+}
+
+impl utimbuf_t {
+    pub fn atime(&self) -> time_t {
+        self.atime
+    }
+
+    pub fn mtime(&self) -> time_t {
+        self.mtime
+    }
+}
+
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+pub enum Utime {
+    UTIME_OMIT,
+    UTIME_NOW,
+    UTIME(Timespec),
+}
+
+pub fn get_utimes(times: Option<(timespec_t, timespec_t)>) -> Result<(Utime, Utime)> {
+    let now = OcclumTimeProvider.current_time();
+    if let Some(times) = times {
+        let (atime, mtime) = times;
+
+        if (!timespec_valid(atime)) || (!timespec_valid(mtime)) {
+            return_errno!(EINVAL, "invalid argument: times")
+        }
+
+        let atime = if atime.nsec() == UTIME_OMIT {
+            Utime::UTIME_OMIT
+        } else if atime.nsec() == UTIME_NOW {
+            Utime::UTIME(now)
+        } else {
+            Utime::UTIME(Timespec {
+                sec: atime.sec(),
+                nsec: atime.nsec(),
+            })
+        };
+        let mtime = if mtime.nsec() == UTIME_OMIT {
+            Utime::UTIME_OMIT
+        } else if mtime.nsec() == UTIME_NOW {
+            Utime::UTIME(now)
+        } else {
+            Utime::UTIME(Timespec {
+                sec: mtime.sec(),
+                nsec: mtime.nsec(),
+            })
+        };
+        Ok((atime, mtime))
+    } else {
+        Ok((Utime::UTIME(now), Utime::UTIME(now)))
+    }
+}
+
+fn timespec_valid(time: timespec_t) -> bool {
+    if (time.nsec() == UTIME_NOW || time.nsec() == UTIME_OMIT) {
+        true
+    } else {
+        time.sec() >= 0 && time.nsec() >= 0 && time.nsec() < 1_000_000_000
+    }
+}
+
+pub fn do_utimes(
+    dirfd: i32,
```

Comment:
Is it better to use FileDesc, FsPath as the parameter of the functions in utimes.rs.
You could add a do_utimes function in syscalls.rs to do the translation, or just move this function to syscalls.rs.

Reply:
I add `do_utimes_wrapper()` in `syscalls.rs` since `do_utimes` is already used.